k3k-kubelet/kubelet.go

```diff
@@ -0,0 +1,274 @@
+	if err != nil {
+		return nil, err
+	}
+
+	virtClient, err := kubernetes.NewForConfig(virtConfig)
+	if err != nil {
+		return nil, err
+	}
+	return &kubelet{
+		name:       c.NodeName,
+		hostConfig: hostConfig,
+		hostClient: hostClient,
+		virtClient: virtClient,
+	}, nil
+}
+
+func (k *kubelet) RegisterNode(srvPort, namespace, name, hostname string) error {
+	providerFunc := k.newProviderFunc(namespace, name, hostname)
+	nodeOpts := k.nodeOpts(srvPort, namespace, name, hostname)
+
+	var err error
+	k.node, err = nodeutil.NewNode(k.name, providerFunc, nodeutil.WithClient(k.virtClient), nodeOpts)
+	if err != nil {
+		return fmt.Errorf("unable to start kubelet: %v", err)
+	}
+	return nil
+}
+
+func (k *kubelet) Start(ctx context.Context) {
+	go func() {
+		ctx := context.Background()
+		logger, err := zap.NewProduction()
+		if err != nil {
+			fmt.Printf("unable to create logger: %s", err.Error())
+			os.Exit(1)
+		}
+		wrapped := LogWrapper{
+			*logger.Sugar(),
+		}
+		ctx = log.WithLogger(ctx, &wrapped)
+		err = k.node.Run(ctx)
+		if err != nil {
+			fmt.Printf("node errored when running: %s \n", err.Error())
+			os.Exit(1)
+		}
+	}()
+	if err := k.node.WaitReady(context.Background(), time.Minute*1); err != nil {
+		fmt.Printf("node was not ready within timeout of 1 minute: %s \n", err.Error())
+		os.Exit(1)
+	}
+	<-k.node.Done()
+	if err := k.node.Err(); err != nil {
+		fmt.Printf("node stopped with an error: %s \n", err.Error())
+		os.Exit(1)
+	}
+	fmt.Print("node exited without an error")
+}
+
+func (k *kubelet) newProviderFunc(namespace, name, hostname string) nodeutil.NewProviderFunc {
+	return func(pc nodeutil.ProviderConfig) (nodeutil.Provider, node.NodeProvider, error) {
+		utilProvider, err := provider.New(*k.hostConfig, namespace, name)
+		if err != nil {
+			return nil, nil, fmt.Errorf("unable to make nodeutil provider %w", err)
+		}
+		nodeProvider := provider.Node{}
+
+		provider.ConfigureNode(pc.Node, hostname, k.port)
+		return utilProvider, &nodeProvider, nil
+	}
+}
+
+func (k *kubelet) nodeOpts(srvPort, namespace, name, hostname string) nodeutil.NodeOpt {
+	return func(c *nodeutil.NodeConfig) error {
+		c.HTTPListenAddr = fmt.Sprintf(":%s", srvPort)
+		// set up the routes
+		mux := http.NewServeMux()
+		if err := nodeutil.AttachProviderRoutes(mux)(c); err != nil {
+			return fmt.Errorf("unable to attach routes: %w", err)
+		}
+		c.Handler = mux
+
+		ctx, cancel := context.WithTimeout(context.Background(), time.Second*10)
+		defer cancel()
+		tlsConfig, err := loadTLSConfig(ctx, k.hostClient, name, namespace, k.name, hostname)
+		if err != nil {
+			return fmt.Errorf("unable to get tls config: %w", err)
+		}
+		c.TLSConfig = tlsConfig
+		return nil
+	}
+}
+
+func virtRestConfig(ctx context.Context, VirtualConfigPath string, hostClient ctrlruntimeclient.Client, clusterName, clusterNamespace string) (*rest.Config, error) {
```

Comment:
```suggestion
func virtRestConfig(ctx context.Context, virtualConfigPath string, hostClient ctrlruntimeclient.Client, clusterName, clusterNamespace string) (*rest.Config, error) {
```

We never need to capitalize a variable name in an argument definition.